File: conformance/bundles/15-custom-parameter-format/money.steps.kt
@file:JvmName("MoneySteps")

// Kotlin sibling of money.steps.ts / money.steps.py / MoneySteps.java
// (bundle 15-custom-parameter-format) — exercises StepsScope.param's
// `format` parameter: the inverse of `parse`, rendering a value back in the
// document's notation. The sensor returns the WRONG Money on purpose: the
// golden pins the formatted actual ("£2.60"), proving every port renders
// parameter mismatches through `format` identically. Without a format this
// actual would be each port's native object rendering, which is deliberately
// outside conformance.
package dev.varar.kotlin.conformance.bundle15

import dev.varar.kotlin.steps
import dev.varar.kotlin.sensor
import java.util.Locale

val steps = steps {
    param(
        "money",
        Regex("""£\d+\.\d{2}"""),
        format = { m: Map<String, Any> -> String.format(Locale.ROOT, "£%.2f", m["value"]) },
    ) { groups ->
        mapOf("currency" to "GBP", "value" to groups[0].substring(1).toDouble())
    }
    sensor("The late fee is {money}") { fee: Map<String, Any> ->
        mapOf("currency" to "GBP", "value" to 2.6)
    }
}

File: conformance/bundles/15-custom-parameter-format/money.steps.cs
// C# sibling of money.steps.ts / .rs (bundle 15-custom-parameter-format).
// Money is encoded as a bare Float (pounds); format renders it back in document notation, so the
// pinned mismatch reads £2.60 / £2.55.
using System.Globalization;
using Varar;
using Varar.Core;

namespace Varar.Corpus.B15;

public static class MoneySteps
{
    public static void Register(Steps s)
    {
        s.Param(
            "money",
            @"£\d+\.\d{2}",
            g =>
            {
                var raw = g[0]!;
                var body = raw.StartsWith("£", StringComparison.Ordinal) ? raw.Substring(1) : raw;
                return Value.Of(double.TryParse(body, NumberStyles.Float, CultureInfo.InvariantCulture, out var d) ? d : 0.0);
            },
            v => v is VFloat f ? "£" + f.Float.ToString("F2", CultureInfo.InvariantCulture) : v.ToString()!);

        // Returns the WRONG money on purpose; the golden pins the formatted actual "£2.60", proving
        // mismatches render through format.
        s.Sensor("The late fee is {money}", (state, expected) => Value.Of(2.6));
    }

    public static Value State() => Value.Null;
}
